include/libtorrent/packet_pool.hpp

```diff
@@ -0,0 +1,180 @@
+		bool need_resend:1;
+
+		// this is set to true for packets that were
+		// sent with the DF bit set (Don't Fragment)
+		bool mtu_probe:1;
+
+	#if TORRENT_USE_ASSERTS
+		int num_fast_resend;
+	#endif
+
+		// the actual packet buffer
+		std::uint8_t buf[1];
+	};
+
+
+	struct packet_deleter
+	{
+		void operator()(packet* handle) const
+		{
+			handle->~packet();
+			std::free(handle);
+		}
+	};
+
+	using packet_ptr = std::unique_ptr<packet, packet_deleter>;
+
+	template<int ALLOCATE_SIZE>
+	struct TORRENT_EXTRA_EXPORT packet_slab
+	{
+		static const int allocate_size{ ALLOCATE_SIZE };
+
+		explicit packet_slab(std::size_t limit) : m_limit(limit) { m_storage.reserve(m_limit); }
+
+		packet_slab(const packet_slab&) = delete;
+
+		void try_push_back(packet_ptr &p)
+		{
+			if (m_storage.size() < m_limit)
+				m_storage.push_back(std::move(p));
+		}
+
+		packet_ptr alloc()
+		{
+			if (m_storage.empty())
+				return packet_ptr(static_cast<packet*>(std::malloc(sizeof(packet) + allocate_size)));
+			else
+			{
+				auto ret = std::move(m_storage.back());
+				m_storage.pop_back();
+				return ret;
+			}
+		}
+
+		void flush() { m_storage.resize(0); }
+	private:
+		const std::size_t m_limit;
+		aux::vector<packet_ptr, std::size_t> m_storage;
+	};
+
+	// single thread packet allocation packet pool
+	// can handle common cases of packet size by 3 pools
+	struct TORRENT_EXTRA_EXPORT packet_pool : private single_threaded
+	{
+		packet *acquire(int allocate)
+		{
+			TORRENT_ASSERT(is_single_thread());
+			TORRENT_ASSERT(allocate >= 0);
+			TORRENT_ASSERT(allocate <= std::numeric_limits<std::uint16_t>::max());
+
+			packet_ptr p{ alloc(allocate) };
+			new (p.get()) packet();
+			p->allocated = static_cast<std::uint16_t>(allocate);
+			return p.release();
+		}
+
+		void release(packet *p)
+		{
+			TORRENT_ASSERT(is_single_thread());
+
+			if (p == nullptr) return;
+
+			packet_ptr pp(p); // takes ownership and may auto free if slab container does not move it
+			std::size_t allocated = pp->allocated;
+
+			if (allocated <= m_syn_slab.allocate_size) { m_syn_slab.try_push_back(pp); }
+			else if (allocated <= m_mtu_floor_slab.allocate_size) { m_mtu_floor_slab.try_push_back(pp); }
+			else if (allocated <= m_mtu_ceiling_slab.allocate_size) { m_mtu_ceiling_slab.try_push_back(pp); }
+		}
+
+		//TODO: call from utp manager to flush mem
+		void flush()
+		{
+			TORRENT_ASSERT(is_single_thread());
+
+			m_syn_slab.flush();
+			m_mtu_floor_slab.flush();
+			m_mtu_ceiling_slab.flush();
+		}
+
+	private:
+		packet_ptr alloc(int allocate)
+		{
+			if (allocate <= m_syn_slab.allocate_size) { return m_syn_slab.alloc(); }
+			else if (allocate <= m_mtu_floor_slab.allocate_size) { return m_mtu_floor_slab.alloc(); }
+			else if (allocate <= m_mtu_ceiling_slab.allocate_size) { return m_mtu_ceiling_slab.alloc(); }
+			return packet_ptr(static_cast<packet*>(std::malloc(sizeof(packet) + allocate)));
+		}
+		packet_slab<sizeof(utp_header)> m_syn_slab{ 0x100 };
```

Comment:
in fact, the size of these is not directly related to how popular a packet size is. You can have thousands of packets allocated, sitting in the send buffer, with a very small slab limit. This size only determines how many to keep around when churn fluctuates. I would think you would get the majority of performance wins by making these be limited to 10 or 20 or so.